src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const clone1 = { ...state };
+  const clonesArray = [];
+
+  for (let key = 0; key < actions.length; key++) {
+    if (actions[key].type === 'addProperties') {
+      Object.assign(clone1, actions[key].extraData);
+    } else if (actions[key].type === 'removeProperties') {
+      for (let i = 0; i < actions[key].keysToRemove.length; i++) {
+        delete clone1[`${actions[key].keysToRemove[i]}`];
+      }
+    } else if (actions[key].type === 'clear') {
+      for (const j in clone1) {
+        delete clone1[j];
+      }
+    }
```

Comment:
```suggestion
  for (let key = 0; key < actions.length; key++) {
    if (actions[key].type === 'addProperties') {
      Object.assign(clone1, actions[key].extraData);
    } else if (actions[key].type === 'removeProperties') {
      for (let i = 0; i < actions[key].keysToRemove.length; i++) {
        delete clone1[`${actions[key].keysToRemove[i]}`];
      }
    } else if (actions[key].type === 'clear') {
      for (const j in clone1) {
        delete clone1[j];
      }
    }
```
You can iterate through array of actions with FOR OF, code will look simpler.

Reply:
hmm, yes indeed, thanks)
but not FOR OFF, but FOR IN =)

Reply:
actions is an array, if I'm not mistaken, with arrays we should use for of, but if it works with in, then it's ok ;))